### ISSUE TITLE
Internal APIs currently needed for several significant use-cases

### ISSUE THREAD

**Issue by symbioquine** [isomorphic-git/isomorphic-git#1071]
First, thanks for the creating this awesome library!

There are a few use-cases which seem to be difficult to implement without access to APIs that are currently designated as "internal".

### Decorated HTTP plugins

The decorator pattern would be a natural choice for those wishing to modify the default http behavior in new ways without completely re-implementing it. However, as far as I can tell, once you replace the default http plugin, there isn't a good way to delegate to it from the decorator.

### Implementing the server-side of pack upload/download

Most of the pieces already exist in isomorphic-git for a client to use the library to implement a git server supporting pushing/pulling changes. I have a prototype which supports git clone over WebRTC which required very little glue code - the main new piece was an 'side-band-64k' protocol implementation based on the commented code already present in isomorphic-git.

Unfortunately, this already required access to several internal parts of isomorphic-git (`uploadPack`, `GitPktLine`, `collect`, `parseUploadPackRequest`, `listObjects`), without which I would have had to duplicate a great deal more code.

**Comment by wmhilton** [comment]
> ### Decorated HTTP plugins
> 
> The decorator pattern would be a natural choice for those wishing to modify the default http behavior in new ways without completely re-implementing it. However, as far as I can tell, once you replace the default http plugin, there isn't a good way to delegate to it from the decorator.

Actually, since v1.0 does not have any plugins, using a decorator with the http clients is easy. 😀 The node and web HTTP clients are exported separately; you can wrap its `request` method and pass the modified client to `git.clone`, etc. See the [docs on `http` client argument](https://isomorphic-git.org/docs/en/http)

> ### Implementing the server-side of pack upload/download
> 
> Most of the pieces already exist in isomorphic-git for a client to use the library to implement a git server supporting pushing/pulling changes. I have a prototype which supports git clone over WebRTC which required very little glue code - the main new piece was an 'side-band-64k' protocol implementation based on the commented code already present in isomorphic-git.
> 
> Unfortunately, this already required access to several internal parts of isomorphic-git (`uploadPack`, `GitPktLine`, `collect`, `parseUploadPackRequest`, `listObjects`), without which I would have had to duplicate a great deal more code.

HOLY CRAP THAT'S AWESOME! ❤️ ❤️ ❤️ 

I'm extremely interested in working on the server-side of the equation! (And _particularly_ this WebRTC use case... SO EXCITED you have a prototype?!!) There's a [feat/server](https://github.com/isomorphic-git/isomorphic-git/tree/feat/server) branch (based on v0.x) I was working on in December, along with a repo https://github.com/isomorphic-git/server. I think the main addition I'd made was `indexPack` returned a list of all the objects in the packfile instead of returning nothing, and I'd added some new functions `serveInfoRefs`, `serveReceivePack`, and  `parseReceivePackRequest`. I hadn't quite finished it: my prototype server would let you push to it, and it would verify all the commits' PGP signatures, but it didn't actually save the packfile and update the refs.

The reason I removed `internal-apis` from the published package is I didn't want someone to rely on them and then get screwed if I change the internals. Have you tried building `isomorphic-git` from source? I'd love it if we could collaborate on the server stuff!

I already told @crutchcorn I'd work on getting `isomorphic-git` working in React Native this weekend, but I'd love to help you out too! Maybe I can do both at once... anyhow. Yeah! Any chance I can peek at your prototype to see how you did it? Then I can combine the ideas from your code and from the unfinished https://github.com/isomorphic-git/server and figure out what the ideal public API functions would be. 🤔 And if we're not 100% sure, we can do what I did for `walk` and just stick `-Beta1` at the end of the function names, and so then we ought to have a workable `isomorphic-git` for servers API within a couple weeks!

**Comment by crutchcorn** [comment]
Speaking of, is there a Discord or Slack of some kind associated with the project? It may help with scenarios like this where more long-form real-time conversation could be had that could then be mirrored back to GH.

**Comment by wmhilton** [comment]
There's a Gitter... https://gitter.im/isomorphic-git That was hip in 2017. 😂

**Comment by wmhilton** [comment]
Although the notifications in the mobile app keep stopping working. Maybe if we used an app like Slack that had working notifications I'd reply more often. 🤔

**Comment by symbioquine** [comment]
@wmhilton Thanks for the quick response and the enthusiasm.

That's great to hear about 1.0 simplifying the http decoration use-case!

I've uploaded the prototype code at https://github.com/symbioquine/git-webrtc-prototype Please bear with the seriously hacky prototype code :)

## Abbreviated walk-through of key pieces of the prototype

### Smart protocol handling

https://github.com/symbioquine/git-webrtc-prototype/blob/master/src/index.js#L50-L87

### Partial 'side-band-64k' implementation - doesn't support progress or errors

https://github.com/symbioquine/git-webrtc-prototype/blob/master/src/IsoGitMuxSideBand.js#L64-L94

### Request-reply wrapper over PeerJS

https://github.com/symbioquine/git-webrtc-prototype/blob/master/src/WebRTCPeerNet.js#L6-L152

### Decorated "hybrid" HTTP plugin

https://github.com/symbioquine/git-webrtc-prototype/blob/master/src/index.js#L118-L152

*Note: this could probably be replaced with whatever comes out of https://github.com/isomorphic-git/isomorphic-git/issues/97*

### Unnecessary and likely dangerous code

*I originally thought I could expect the pack files to already be on disk so these lines are unnecessary and likely dangerous since they could potentially serve arbitrary files in the file-system.*

https://github.com/symbioquine/git-webrtc-prototype/blob/master/src/index.js#L89-L93

**Comment by wmhilton** [comment]
This is still at the forefront of my Todo List... I've just been too busy or too tired. And a little stressed about this coronavirus thing.

**Comment by wmhilton** [comment]
@symbioquine I read your [WebRTCPeerNet.js](https://github.com/symbioquine/git-webrtc-prototype/blob/master/src/WebRTCPeerNet.js) code and was concerned that the entire packfile was sent as one message. I thought before we went any further, we should make sure we can using a streaming protocol.

So this is what I've been working on: https://github.com/wmhilton/lightning-http It's a pure JS implementation of HTTP/1.1 chunked transfer-encoding\*. To prove out the concept, I implemented a backend that uses Node's raw `net` module and a very simple Express-like routing helper. After a little more refactoring, the only things the backend will need to do is establish the connection and provide a `write` method for writing binary data and a `data` event for receiving binary data. This should make it trivial to write WebSocket and WebRTC backends, and maybe even a ServiceWorker backend.

\* fun fact, my parser was inspired _yet again_ by Tim Caswell's prescient work: https://github.com/creationix/http-parser-js

**Comment by symbioquine** [comment]
> I read your WebRTCPeerNet.js code and was concerned that the entire packfile was sent as one message. I thought before we went any further, we should make sure we can using a streaming protocol.

Yeah, for my prototype I was just relying on the browser does its own MTU negotiation/chunking under the hood with WebRTC, but pulling that stuff up into the library itself would allow for controlling progress/cancellation/timeouts more explicitly.

Overall your approach sounds very cool! Can't wait to look under the hood of your proof-of-concept...

**Comment by robert-cronin** [comment]
has there been any update on this? I'd love access to internal apis for building a custom backend but I think server capabilities exposed in the public api would negate the need for that! Would be a great feature to have.

**Comment by dionysiusmarquis** [comment]
Git request/response tunneled as stream would be heartwarming. I was just looking for a way to implement decentralized git for `libp2p`

**Comment by hrueger** [comment]
I'm also interested in this, specifically `uploadPack`. I'm currently trying to make https://gabrielcsapo/node-git-server use isomorphic-git instead of system git calls.

Is there any ETA?

**Comment by jcubic** [comment]
@hrueger this project is driven by community. It doesn't have main author that do free labor. If you want this feature, consider contributing. If you don't want, don't expect this feature to ever be implemented, unless you can find someone else that will do this for you.

**Comment by hrueger** [comment]
Sorry, I didn't want to sound rude or demanding. I was just wondering - maybe someone already implemented that in a fork or something. Definitely looking into contributing.

**Comment by jcubic** [comment]
You don't need to apologize, the PR are merged pretty fast, so if someone have a feature implemented it's included with the main repo, and releases are automated. I only need to merge, which I can do even on my phone, which I'm using right now.

If you want to contribute that would be great, you are free to implement the feature as you want we only need to agree on the API, that need to match to the rest of the project.

**Comment by hrueger** [comment]
Hi @jcubic,
I'm currently fiddling around with this. I find a lot of calls (like here, here and here) to 
```TypeScript
await file.read(`${gitdir}/someFile`, { encoding: 'utf8'})
```
Is this supposed to work somehow? According to the [Node Docs](https://nodejs.org/api/fs.html#fsreadfd-options-callback) `fs.read` takes a filehandle, not a string and a callback, not options.

Is there some kind of file system abstraction which I've overlooked?

**Comment by jcubic** [comment]
The link your provided is a promisified API in Node `fs.promise.readFile`. And I think that the code for `readFile` also accepts "utf8" like the NodeJS version.

**Comment by hrueger** [comment]
Thanks for the quick response. However, I don't completely understand.

> The link your provided is a promisified API in Node fs.promise.readFile

Maybe it does? However, all the `read()` (not `readFile`) methods (be it sync, async with callback or async with promises) take a file handle as the first parameter and not a string...

Edit: Never mind. I just found that all the Methods which are exposed wrap `fs` in the `FileSystem` class. That means, when accessing `internal-api`, I need to do this manually.

Sorry for the noise
